File: app/src/main/java/bible/translationtools/converter/ConverterTask.java
package bible.translationtools.converter;

import android.os.AsyncTask;
import androidx.fragment.app.Fragment;

/**
 * Created by mXaln on 11/11/17.
 */

public final class ConverterTask extends AsyncTask<Void, Integer, Integer> {

    ConverterResultCallback callback;

    public ConverterTask(Fragment c) {
        this.callback = (ConverterResultCallback) c;
    }

    @Override
    protected Integer doInBackground(Void... params) {
        return callback.startConversion();
    }

    @Override
    protected void onPreExecute() {
        callback.conversionStarted();
        super.onPreExecute();
    }

    @Override
    protected void onProgressUpdate(Integer... values) {
        super.onProgressUpdate(values);
    }

    @Override
    protected void onPostExecute(Integer result) {
        callback.conversionDone(result);
        super.onPostExecute(result);
    }

    public interface ConverterResultCallback {
        Integer startConversion();
        Void conversionStarted();
        Void conversionDone(Integer result);
    }
}

File: app/src/main/java/bible/translationtools/converter/ConverterTask.kt
package bible.translationtools.converter

import android.os.AsyncTask
import androidx.fragment.app.Fragment

/**
 * Created by mXaln on 11/11/17.
 */
class ConverterTask(c: Fragment?) : AsyncTask<Void?, Int?, Int?>() {
    var callback: ConverterResultCallback

    init {
        this.callback = c as ConverterResultCallback
    }

    override fun doInBackground(vararg params: Void?): Int {
        return callback.startConversion()
    }

    override fun onPreExecute() {
        callback.conversionStarted()
        super.onPreExecute()
    }

    override fun onProgressUpdate(vararg values: Int?) {
        super.onProgressUpdate(*values)
    }

    override fun onPostExecute(result: Int?) {
        callback.conversionDone(result)
        super.onPostExecute(result)
    }

    interface ConverterResultCallback {
        fun startConversion(): Int
        fun conversionStarted()
        fun conversionDone(result: Int?)
    }
}
